### DISCUSSION TITLE
How to redirect `www.example.com` to `example.com`?

### DISCUSSION THREAD

**Discussion by Glitchy-Tozier** [rwf2/Rocket#2336] [Questions] [answered]
I already got this code to redirect to pages without trailing slashes.
However, I have absolutely no clue how to make change this handler to also redirect to non-`www.`-pages.

```rs
/// A handler that redirects to the no-trailing-slash version
/// of the URL, if the URL ends in a trailing slash.
#[derive(Clone)]
pub struct SlashRemover;
#[rocket::async_trait]
impl Handler for SlashRemover {
    async fn handle<'r>(&self, req: &'r Request<'_>, data: Data<'r>) -> Outcome<'r> {
        if req.uri().path().ends_with('/') && req.uri().path().to_string().chars().count() > 1 {
            println!("{}", req.uri().clone().into_owned().into_normalized());
            Outcome::from(
                req,
                Redirect::to(req.uri().clone().into_owned().into_normalized()),
            )
        } else {
            Outcome::forward(data)
        }
    }
}
impl From<SlashRemover> for Vec<Route> {
    fn from(sr: SlashRemover) -> Vec<Route> {
        vec![Route::new(Method::Get, "/<..>", sr)]
    }
}
```

**Comment by the10thWiz** [comment, marked as answer]
This is actually fairly simple. First, you need to detect that the client has made a request for `www.ex.com` rather than `ex.com`, which can be checked via the `Host` header. [`Request`](https://api.rocket.rs/v0.5-rc/rocket/struct.Request.html#method.host) has a specific method for checking the host (note that if the Host is not sent, you CANNOT identify what domain the user-agent is actually requesting). Then, simply redirect to the correct domain, via the same `Redirect` method. I recommend using the [`uri!()`](https://api.rocket.rs/v0.5-rc/rocket/macro.uri.html) macro if possible.

However, I'm also curious what the utility of this is. Rocket's default behavior is to ignore the trailing slash, and by default, unless you create specific Fairings to differentiate the various subdomains, Rocket will ignore the subdomain. The only potential issue I could see cropping up is a subdomains might have separate cookies & local storage on the client side. This can be solved for cookies by setting the `Domain` attribute to `ex.com`, which instructs the browser to only send the cookies to `ex.com`, and any subdomains, such as `www.ex.com`.

**Reply by Glitchy-Tozier** [reply]
Thank you for your response, I'll play around with that.

>However, I'm also curious what the utility of this is.

I wanted to do this because
1. Apparently this is a good practice for SEO (that is, pick only www-pages or only non-www-pages)
2. My website often uses invite-links that I try to keep as short as possible.

**Reply by the10thWiz** [reply]
In this case, you should be able to make all internal links use the non-www version, and it should be pretty easy to always make the invite links non-www.

However, I'm not familiar with SEO, so I can't comment on whether this is a good idea or not. I can see arguments both ways, and it really depends on how search engines actually index sites. I hope my advice was able to help, let me know if you run into any issues!
